### PR TITLE
Define `isless` and `isunordered` for `nothing`

### PR DESCRIPTION
**Pull request by andyferris**
Currently `isequal` and `hash` are both well-defined for `nothing`, but not `isless`. This means that we can add `nothing` to a hashmap-based set or dictionary - but not a b-tree-based set or dictionary. I feel this is inconsistent.

This PR provides basic definitions for `isless` so that it takes place in the total ordering of objects. Like `NaN` and `missing` it is ordered "greater than" ordinary types. In this PR it is ordered such that `isless(missing, nothing)` is `true` (I made `nothing` greatest because it is somehow more fundamental to Julia than `missing`). This came up for me because we had collections of type `Union{T, Nothing}` for various `T` in our application code, and adding these definitions seemed like a no-brainer.

One thing I haven't done (in the spirit of minimalism of this first cut) is define any particular behavior for `<`, etc, so it is following the generic behavior based on `isless`. For instance you can get `0 < nothing` being `true`. If we wanted to make that error out, what error type should it be? Or would we want it to behave like `NaN`, so that `0 < nothing` is `false` (as well as `nothing < 0` is false)? Or should there be some other behavior? (I really don't think we want to start "propagating" `nothing`s).

### CODE REVIEW

File: base/missing.jl [hunk 1]
```diff
@@ -86,6 +86,8 @@ isequal(::Any, ::Missing) = false
 isless(::Missing, ::Missing) = false
 isless(::Missing, ::Any) = false
 isless(::Any, ::Missing) = true
+isless(::Missing, ::Nothing) = true
```
Comment (quinnj) [inline base/missing.jl:89]: Shouldn't this method be covered by `isless(::Any, ::Nothing)`?

**Comment by quinnj** [comment]
There are other cases where we specifically throw a `MethodError` (like [here](https://github.com/JuliaLang/julia/blob/045d10e371c222000d4386240694f71a89383f8e/base/array.jl#L533)), so I'd prefer to keep that as-is for `<`. Otherwise, I think this makes sense to me; we're saying `isequal`, `hash`, and `isless` are generic "object" methods that should work on anything.

**Comment by JeffBezanson** [comment]
I do think `<` should continue to be an error. But it's also never been the case that `isless` was expected to work on everything; only objects with a total order

**Comment by andyferris** [comment]
Awesome, thanks, that's what I guessed would be wanted. I've implemented the `MethodError`s now and added NEWS.

> But it's also never been the case that `isless` was expected to work on everything; only objects with a total order

That's fair enough.

I suppose my point of view is that objects at the end of the day are bits; there is always at least one _possible_ total order (e.g. for reference types we even hash pointers; you could similarly order a sorted-tree by pointers if needed). I tend to add `isless` to my types at the same time as `hash` because, to me, they are there to provide the same utility (automatic dictionary and set support).

File: base/missing.jl (hunk 1 above)
Comment (andyferris) [reply base/missing.jl:91]: It's here to resolve a method ambiguity between `isless(::Missing, ::Any)` and `isless(::Any, ::Nothing)`.

**Comment by andyferris** [comment]
The sole CI error regards an unrelated FreeBSD distributed test

**Comment by cmcaine** [comment]
This PR might need to add `isunordered(x::Nothing) = true` so that `Base.isgreater` (used in `findmin`) keeps its relationship with `isless`.

Some docstrings that specify that `Missing` is the largest/smallest value should probably be updated, too. There are docstrings like this in base/operators.jl, base/reduce.jl and base/reducedim.jl, and perhaps elsewhere, too.

**Comment by andyferris** [comment]
Thanks @cmcaine - I've done that.

**Comment by andyferris** [comment]
I found a few extra references in the manual to `missing` being the greatest according to `isless` and have adjusted those too. (I did an exhaustive search so I hope this should be all of them).

**Comment by JeffBezanson** [comment]
:-1: from triage.

**Comment by quinnj** [comment]
It'd be helpful to those not able to attend the triage call to walk through the reasoning for 👎 and closing; cc: @JeffBezanson @vtjnash

**Comment by andyferris** [comment]
I agree - it would be good to understand the reasoning.

(As it is, for me this change would solve a real-world issue at work as well as support some open-source efforts with sort-order-based dictionaries and lookups, so I would like to understand what was considered negative about this change, even if only to avoid problems elsewhere).

**Comment by StefanKarpinski** [comment]
The gut reaction against this was just that it's adding yet another behavior for `nothing`. I was focused on the `<` stuff but the `isless` part is potentially justified. For example, I do not think that we should have `0 < nothing` be false or true, it really should to be a method error as it is now. I'm also unclear on the motivation for `isless`: why are people putting `nothing` in ordered collections in the first place? I get that it's not consistent with hashes, but in general `isless` isn't a total order on all objects, so that's generally the case. We would need `isless` to be a total order for order-based collections to be as general as hash-based collections, and that's something we've avoided so far.

**Comment by StefanKarpinski** [comment]
I guess there's a different conversation to be had about how to support order-based collections that might hold heterogenous values. One essentially wants `isless` but made total.

**Comment by andyferris** [comment]
Yes, I think that is the crux of the matter.

In my case it was (some more complicated variation of) parsing JSON and dealing with that data. The approach taken so far has been to propagate JSON's `null` (or optional fields) as `nothing` - in these places `missing` was unsuitable because we wanted to avoid accidental propagation so it has worked well so far (like how `nothing` works well for `iterate`). As the pattern continued, in various places we ended up with values of `Union{T, Nothing}`. To be clear, we've been careful not to assign behavior for `nothing` (it's always checked for via `isnothing` or whatever) but we do sometimes want to e.g. `sort(::Vector{Union{T, Nothing}})` (which is an array you could get straight out of JSON). We're also moving towards using sorted sets and dictionaries with flexible key/value types for various reasons.

For us the alternative would be to create our own non-propagating hybrid of `nothing` and `missing`, parse inputs into that, and provide the same definitions as here. (Or perhaps switch to `missing` and accept we might experience some undetected bugs if we are careless in places). Given heterogenous order-based collections is a thing other people might want, I thought this contribution was the simplest approach.

> For example, I do not think that we should have `0 < nothing` be `false` or `true`, it should continue to be a method error.

Yes, I think we all agree with that (as was implemented here).

**Pull request closed without merging**